src/transformers/models/aria/modular_aria.py

```diff
@@ -0,0 +1,1380 @@
+        **loss_kwargs,
+    ) -> Union[Tuple, AriaCausalLMOutputWithPast]:
+        """
+        Forward pass of the `AriaForConditionalGeneration` model.
+
+        This method processes both text and image inputs, merges them if necessary,
+        and generates output using the language model.
+
+        Args:
+            input_ids (`torch.LongTensor`, *optional*):
+                Input token IDs.
+            pixel_values (`torch.FloatTensor`, *optional*):
+                Pixel values of the images.
+            pixel_mask (`torch.LongTensor`, *optional*):
+                Mask for the pixel values.
+            attention_mask (`torch.Tensor`, *optional*):
+                Attention mask.
+            position_ids (`torch.LongTensor`, *optional*):
+                Position IDs.
+            past_key_values (`List[torch.FloatTensor]`, *optional*):
+                Past key values for efficient processing.
+            inputs_embeds (`torch.FloatTensor`, *optional*):
+                Input embeddings.
+            labels (`torch.LongTensor`, *optional*):
+                Labels for computing the language modeling loss.
+            use_cache (`bool`, *optional*):
+                Whether to use the model's cache mechanism.
+            output_attentions (`bool`, *optional*):
+                Whether to output attention weights.
+            output_hidden_states (`bool`, *optional*):
+                Whether to output hidden states.
+            return_dict (`bool`, *optional*):
+                Whether to return a `ModelOutput` object.
+            num_logits_to_keep (`int`, *optional*, defaults to 0):
+                Calculate logits for the last `num_logits_to_keep` tokens, or all `input_ids` if `0`.
+            cache_position (`torch.LongTensor`, *optional*):
+                Cache positions.
+            **loss_kwargs:
+                Additional keyword arguments for loss calculation.
+
+        Returns:
+            `Union[Tuple, AriaCausalLMOutputWithPast]`:
+                Model outputs.
+        """
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if inputs_embeds is None:
+            inputs_embeds = self.get_input_embeddings()(input_ids)
+
+        # 2. Merge text and images
+        if pixel_values is not None and inputs_embeds.shape[1] != 1:
+            if input_ids is None:
+                special_image_mask = inputs_embeds == self.get_input_embeddings()(
+                    torch.tensor(self.config.image_token_index, dtype=torch.long, device=inputs_embeds.device)
+                )
+                n_image_tokens = (special_image_mask).sum(dim=1)[0][0].item()
+            else:
```

Comment:
This was done to pass the tests that require input_embeds only! 
It's more like a new feature

Reply:
ohh, I see, tbh most VLMs now raise an error if `inputs_embeds` is passed along with `pixel_values`, thus we overwrote the tests. For me Aria solutions seems to be better, and should work in most cases right? Or can it be that the tensors will get small numerical diff because of casting to `dtypes`, and the hard `==` might not work?